### PR TITLE
rgw/includes: Clang does not like the current lincludes for the rgw modules

### PR DESCRIPTION
**Pull request by wjwithagen**
Changed the include order in several rgw files, to please Clang.

### CODE REVIEW

**Comment by cbodley** [comment]
Hi @wjwithagen, I'm to blame for this RGWPeriodHistory stuff. You're right that std::deque needs a complete type; I'm surprised that gcc let me get away with it.

I think a better solution would be to just have rgw_period_history.h include rgw_rados.h, so that we don't have to worry about include order everywhere. Otherwise, it would be easy for us gcc folks to mess it up again in the future.

**Comment by wjwithagen** [comment]
@cbodley 
The reason I actually split them apart is that if you include the Rados headers the wrong way, it can happen that you make an inclusion loop whereby the the type is only forward declared,  aut the actual structure is left open, and then there are again more errors of this nature.
I tried it the other way:
    include rgw_period_history in rgw_rados.h, and that definitely falls in the above trap.
Lets see if including rgw_rados.h in gw_period_history.h does not bomb out as well.

**Comment by cbodley** [comment]
@wjwithagen There shouldn't be a problem with circular dependencies here. rgw_period_history.h depends on rgw_rados.h for the definition of RGWPeriod, but rgw_rados.h only needs a forward declaration of RGWPeriodHistory because it wraps it in a std::unique_ptr

**Comment by wjwithagen** [comment]
@cbodley 
And as a side remark.... I've disabled quite a lot of warnings ATM in Clang because otherwise there are a lot of warnings and notices. But in the end it would be nice/good to turn them on and make Clang happy as well. I've got the idea that it is a lot pickier compared to GCC.

**Comment by cbodley** [comment]
@wjwithagen Agreed, and making clang a part of ceph's CI (and turning on things like static analysis and sanitizers) could go a long way to improving code quality. Have you looked into adding a gitbuilder or jenkins for clang/freebsd builds?

**Comment by wjwithagen** [comment]
@cbodley I'm currently running a fresh compile with rgw_rados. included into rgw_period_history.h
Lets see what Clang thinks.

**Comment by wjwithagen** [comment]
@cbodley 
My idea was to first get ceph to compile AND test, but tests keep getting added and sometimes test that used to work start failing... So I've yet to get to the point where all tests "sort"  of complete successfully. ATM that is even with excluded RBD(linux only) en BlueStore( AIO API not compatible).
Lots of times I'm getting python things that do not do what they are supposed to do.... Like rados.py all of a sudden not begin available. Turns out somebody(tm) hid it behind an 'if RBD' conditional. And figuring those kinds of things out take a lot of time....

So how much of completion do we need to start running either a gitbuilder or a jenkins thingy? I know that FreeBSD tests a lot of her stuff thru Jenkins as well. And I'm able to supply a FreeBSD VM to actually run stuff in.

**Comment by cbodley** [comment]
> So how much of completion do we need to start running either a gitbuilder or a jenkins thingy?

I think it would be valuable to start with one that just runs make (with whatever configuration you have working), to verify that the progress you've made so far remains intact.

adding @alimaredia, who's also interested in clang for CI

**Comment by wjwithagen** [comment]
@cbodley @alimaredia Well it sort runs make (gmake that is) in my environment to completion.
It even runs as far as 110 out of 123 tests in make check.

This is the routine I use for general test runs:

  \#!/bin/sh -xve

        \# sudo ./install-deps.sh
        . ./autogen_freebsd.sh
        ./autogen.sh
        ./configure ${CONFIGURE_FLAGS}
        ( cd src/gmock/gtest; patch < /usr/ports/devel/googletest/files/patch-bsd-defines )
        gmake -j8 ENABLE_GIT_VERSION=OFF
        gmake check ENABLE_GIT_VERSION=OFF CEPH_BUFFER_NO_BENCH=yes

**Comment by wjwithagen** [comment]
@cbodley 
A Clang warning that just flew by:
test/common/test_time.cc:126:13: warning: taking the absolute value of unsigned type 'rep' (aka 'unsigned long') has no effect
      [-Wabsolute-value]
All kinds of stuff like that.

**Comment by cbodley** [comment]
It looks like there's a fix for that one at https://github.com/ceph/ceph/pull/7726

**Comment by wjwithagen** [comment]
@cbodley Including rgw_rados.h into rgw_period_history.h completed compiling on FreeBSD.
In Linux I now need to 'hack' my way beyond the radosgw-admin output change.

**Comment by cbodley** [comment]
> Including rgw_rados.h into rgw_period_history.h completed compiling on FreeBSD

Excellent. If you can split that one fix into a separate pull request, I'd be happy to merge it.

> In Linux I now need to 'hack' my way beyond the radosgw-admin output change.

https://github.com/ceph/ceph/pull/7723 was merged today to fix this one

**Comment by wjwithagen** [comment]
@cbodley I've seen the fix #7723 pass by, but currently rebasing gives me everytime more work, instead of solving things. So I try to focus in getting the tree to compile again before getting even more changes in that I need to fix before it start compiling again.

**Comment by wjwithagen** [comment]
And yes, looks like #7726 gets at the error I just posted.
But then I'm not yet at the level where warnings are carefully examined to see if the compiler still does the right thing. But nice to see that other pulls are doing this already.

**Comment by wjwithagen** [comment]
@cbodley I'll make another run by removing all the currently included rgw_rados.h just in front of rgw_period_history.h. And then stuff it all into a separate pull.

**Comment by mattbenjamin** [comment]
A new pull request?  You should be able to just overwrite this one (push -f).

**Comment by wjwithagen** [comment]
On 22-2-2016 20:33, Matt Benjamin wrote:
> A new pull request? You should be able to just overwrite this one (push -f).

Yup,

And then without overwritting all my other stuff.
I though I started this one branch on wjwithagen/master, but then it
still ended up on the rest of my wip.
But we'll get there.

--WjW

**Comment by wjwithagen** [comment]
@cbodley 
I've cleaned the pull.
Hope this works.

File: src/rgw/rgw_period_history.h [hunk 1]
```diff
@@ -11,9 +11,9 @@
 #include "include/assert.h"
 #include "include/types.h"
 
-namespace bi = boost::intrusive;
+#include "rgw_rados.h"
```
Comment (cbodley) [inline src/rgw/rgw_period_history.h]: I was expecting a pull request where this line was the only change. Do the rest of the source files really need to mess with include order, after this line is added?

**Comment by wjwithagen** [comment]
@cbodley 
Some of the files did not really have enough rgw_*.h includes to make Clang happy.
At certain places rgw_metadata.h was missing, other place it complained about rgw_periodic_puller.h
I'll put that change in my other pull and see what it gives.

**Comment by wjwithagen** [comment]
@cbodley 
Clang is definitly not happy with just a one-line add in rgw_periodic_history.h.
For example:
compiling rgw/rgw_acl.cc  is mainly not happy because it does not include rgw_periodic_history.h
Errors
In file included from rgw/rgw_acl.cc:9:
In file included from ./include/types.h:21:
In file included from ./include/uuid.h:8:
In file included from ./include/encoding.h:290:
/usr/include/c++/v1/deque:912:49: error: invalid application of 'sizeof' to an incomplete type 'value_type' (aka 'RGWPeriod')
    static const difference_type __block_size = sizeof(value_type) < 256 ? 4096 / sizeof(value_type) : 16;
                                                ^~~~~~~~~~~~~~~~~~
/usr/include/c++/v1/deque:1159:15: note: in instantiation of template class 'std::__1::__deque_base<RGWPeriod,
      std::__1::allocator<RGWPeriod> >' requested here
    : private __deque_base<_Tp, _Allocator>
              ^
rgw/rgw_period_history.h:28:27: note: in instantiation of template class 'std::__1::deque<RGWPeriod, std::__1::allocator<RGWPeriod>
      >' requested here
    std::deque<RGWPeriod> periods;
                          ^
rgw/rgw_period_history.h:18:7: note: forward declaration of 'RGWPeriod'
class RGWPeriod;
      ^
In file included from rgw/rgw_acl.cc:9:
In file included from ./include/types.h:21:
In file included from ./include/uuid.h:8:
In file included from ./include/encoding.h:290:
/usr/include/c++/v1/deque:1169:30: error: '__alloc_traits' is a protected member of 'std::__1::__deque_base<RGWPeriod,
      std::__1::allocator<RGWPeriod> >'
    typedef typename __base::__alloc_traits        __alloc_traits;
                             ^
rgw/rgw_period_history.h:28:27: note: in instantiation of template class 'std::__1::deque<RGWPeriod, std::__1::allocator<RGWPeriod>
      >' requested here
    std::deque<RGWPeriod> periods;
                          ^
/usr/include/c++/v1/deque:904:54: note: declared protected here
    typedef allocator_traits<allocator_type>         __alloc_traits;
                                                     ^
/usr/include/c++/v1/deque:1170:30: error: 'reference' is a protected member of 'std::__1::__deque_base<RGWPeriod,
      std::__1::allocator<RGWPeriod> >'
    typedef typename __base::reference             reference;
                             ^
/usr/include/c++/v1/deque:905:54: note: declared protected here
    typedef value_type&                              reference;
                                                     ^
/usr/include/c++/v1/deque:1171:30: error: 'const_reference' is a protected member of 'std::__1::__deque_base<RGWPeriod,
      std::__1::allocator<RGWPeriod> >'
    typedef typename __base::const_reference       const_reference;
                             ^
/usr/include/c++/v1/deque:906:54: note: declared protected here
    typedef const value_type&                        const_reference;
                                                     ^
/usr/include/c++/v1/deque:1172:30: error: 'iterator' is a protected member of 'std::__1::__deque_base<RGWPeriod,
      std::__1::allocator<RGWPeriod> >'
    typedef typename __base::iterator              iterator;
                             ^
/usr/include/c++/v1/deque:922:64: note: declared protected here
                             difference_type, __block_size>    iterator;
                                                               ^
/usr/include/c++/v1/deque:1173:30: error: 'const_iterator' is a protected member of 'std::__1::__deque_base<RGWPeriod,
      std::__1::allocator<RGWPeriod> >'
    typedef typename __base::const_iterator        const_iterator;
                             ^
/usr/include/c++/v1/deque:924:64: note: declared protected here
                             difference_type, __block_size>    const_iterator;
                                                               ^
/usr/include/c++/v1/deque:1174:30: error: 'size_type' is a protected member of 'std::__1::__deque_base<RGWPeriod,
      std::__1::allocator<RGWPeriod> >'
    typedef typename __base::size_type             size_type;
                             ^
/usr/include/c++/v1/deque:907:54: note: declared protected here
    typedef typename __alloc_traits::size_type       size_type;
                                                     ^
/usr/include/c++/v1/deque:1175:30: error: 'difference_type' is a protected member of 'std::__1::__deque_base<RGWPeriod,
      std::__1::allocator<RGWPeriod> >'
    typedef typename __base::difference_type       difference_type;
                             ^
/usr/include/c++/v1/deque:908:54: note: declared protected here
    typedef typename __alloc_traits::difference_type difference_type;
                                                     ^
/usr/include/c++/v1/deque:1177:30: error: 'pointer' is a protected member of 'std::__1::__deque_base<RGWPeriod,
      std::__1::allocator<RGWPeriod> >'
    typedef typename __base::pointer               pointer;
                             ^
/usr/include/c++/v1/deque:909:54: note: declared protected here
    typedef typename __alloc_traits::pointer         pointer;
                                                     ^
/usr/include/c++/v1/deque:1178:30: error: 'const_pointer' is a protected member of 'std::__1::__deque_base<RGWPeriod,
      std::__1::allocator<RGWPeriod> >'
    typedef typename __base::const_pointer         const_pointer;
                             ^
/usr/include/c++/v1/deque:910:54: note: declared protected here
    typedef typename __alloc_traits::const_pointer   const_pointer;
                                                     ^
/usr/include/c++/v1/deque:1343:30: error: '__map_const_pointer' is a protected member of 'std::__1::__deque_base<RGWPeriod,
      std::__1::allocator<RGWPeriod> >'
    typedef typename __base::__map_const_pointer __map_const_pointer;
                             ^
/usr/include/c++/v1/deque:918:81: note: declared protected here
    typedef typename allocator_traits<__const_pointer_allocator>::const_pointer __map_const_pointer;
                                                                                ^

**Comment by cbodley** [comment]
@wjwithagen ok, thanks for posting the output. it looks like there really is a circular dependency:
rgw_rados.h -> rgw_metadata.h -> rgw_period_history.h -> rgw_rados.h

i think the best way forward is to refactor the rgw_period_history.h header so it doesn't depend on RGWPeriod. i'm testing this change now, and i'll open a pr once it's ready

**Comment by cbodley** [comment]
pr https://github.com/ceph/ceph/pull/7809 moves the std::deque<RGWPeriod> out of the rgw_period_history.h header to break the circular dependency

**Comment by wjwithagen** [comment]
Commited in a different pull: #7809

**Pull request closed without merging**